src/main/java/baritone/pathing/movement/MovementHelper.java

```diff
@@ -293,55 +339,78 @@ static boolean avoidWalkingInto(Block block) {
-            return true;
+            return Optional.of(true);
         }
         if (block == Blocks.LADDER || (block == Blocks.VINE && Baritone.settings().allowVines.value)) { // TODO reconsider this
-            return true;
+            return Optional.of(true);
         }
         if (block == Blocks.FARMLAND || block == Blocks.GRASS_PATH) {
-            return true;
+            return Optional.of(true);
         }
         if (block == Blocks.ENDER_CHEST || block == Blocks.CHEST || block == Blocks.TRAPPED_CHEST) {
-            return true;
+            return Optional.of(true);
         }
+        if (isWater(block)) {
+            return Optional.empty();
+        }
+        if (Baritone.settings().assumeWalkOnLava.value && MovementHelper.isLava(block)) {
+            return Optional.empty();
+        }
+
+        if (block instanceof BlockSlab) {
+            if (!Baritone.settings().allowWalkOnBottomSlab.value) {
+                if (((BlockSlab) block).isDouble()) {
+                    return Optional.of(true);
+                }
+                return Optional.of(state.getValue(BlockSlab.HALF) != BlockSlab.EnumBlockHalf.BOTTOM);
+            }
+            return Optional.of(true);
+        }
+
+        return Optional.of(block instanceof BlockStairs);
+    }
+
+    static boolean canWalkOnPosition(BlockStateInterface bsi, int x, int y, int z, IBlockState state) {
+        Block block = state.getBlock();
         if (isWater(block)) {
             // since this is called literally millions of times per second, the benefit of not allocating millions of useless "pos.up()"
-            // BlockPos s that we'd just garbage collect immediately is actually noticeable. I don't even think its a decrease in readability
+            // BlockPos s that we'd just garbage collect immediately is actually noticeable. I don't even think it's a decrease in readability
             Block up = bsi.get0(x, y + 1, z).getBlock();
             if (up == Blocks.WATERLILY || up == Blocks.CARPET) {
                 return true;
             }
-            if (isFlowing(x, y, z, state, bsi) || block == Blocks.FLOWING_WATER) {
+            if (MovementHelper.isFlowing(x, y, z, state, bsi) || block == Blocks.FLOWING_WATER) {
                 // the only scenario in which we can walk on flowing water is if it's under still water with jesus off
                 return isWater(up) && !Baritone.settings().assumeWalkOnWater.value;
             }
             // if assumeWalkOnWater is on, we can only walk on water if there isn't water above it
             // if assumeWalkOnWater is off, we can only walk on water if there is water above it
             return isWater(up) ^ Baritone.settings().assumeWalkOnWater.value;
         }
-        if (Baritone.settings().assumeWalkOnLava.value && isLava(block) && !isFlowing(x, y, z, state, bsi)) {
-            return true;
-        }
-        if (block == Blocks.GLASS || block == Blocks.STAINED_GLASS) {
```

Comment:
This check for glass might have slipped through the cracks? Should be added back to the cached variant.

Reply:
It did slip through the cracks, thanks for pointing this out